airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_filter.py

```diff
@@ -36,3 +37,72 @@ def filter_records(
+    :param RecordFilter record_filter: Optional filter to apply after date-time filtering
+    :param PerPartitionCursor per_partition_cursor: Optional Cursor used for mapping cursor value in nested stream_state
+
+    """
+
+    def __init__(
+        self,
+        date_time_based_cursor: DatetimeBasedCursor,
+        record_filter: Optional[RecordFilter],
+        per_partition_cursor: Optional[PerPartitionCursor] = None,
+    ):
+        self._date_time_based_cursor = date_time_based_cursor
+        self._delegate = record_filter
+        self._per_partition_cursor = per_partition_cursor
+
+    @property
+    def _cursor_field(self) -> Union[str, Any]:
+        return self._date_time_based_cursor._cursor_field.eval(self._date_time_based_cursor.config)
+
+    @property
+    def _start_date_from_config(self) -> Union[datetime.datetime, Any]:
+        return self._date_time_based_cursor._start_datetime.get_datetime(self._date_time_based_cursor.config)
+
+    @property
+    def _end_datetime(self) -> Union[datetime.datetime, Any]:
+        return (
+            self._date_time_based_cursor._end_datetime.get_datetime(self._date_time_based_cursor.config)
+            if self._date_time_based_cursor._end_datetime
+            else datetime.datetime.max
+        )
+
+    def filter_records(
+        self,
+        records: Iterable[Mapping[str, Any]],
+        stream_state: StreamState,
+        stream_slice: StreamSlice,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        state_value = self._get_state_value(stream_state, stream_slice)
+        filter_date = self._get_filter_date(state_value)
+        if filter_date:
+            records = (
+                record
+                for record in records
+                if self._end_datetime > self._date_time_based_cursor.parse_date(record[self._cursor_field]) > filter_date
+            )
+        if self._delegate:
+            return self._delegate.filter_records(
+                records=records, stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
+            )
+        return records
+
+    def _get_state_value(self, stream_state: StreamState, stream_slice: StreamSlice) -> Optional[str]:
+        if self._per_partition_cursor:
+            partition_state = self._per_partition_cursor.select_state(stream_slice=stream_slice)
+            return partition_state.get(self._cursor_field) if partition_state else None
+        return stream_state.get(self._cursor_field)
+
+    def _get_filter_date(self, state_value: Optional[str]) -> Optional[datetime.datetime]:
+        start_date_parsed = self._start_date_from_config or None
+        state_date_parsed = self._date_time_based_cursor.parse_date(state_value) if state_value else None
+        return max((x for x in (start_date_parsed, state_date_parsed) if x), default=None)
```

Comment:
I'd always expect a value to be returned here. In what scenarios would we expect the filter date to be `None`?

Reply:
agreed, start_date will be always presented